File: python_manager/lib/binance_interactions.py
from eth_utils import address
import web3, time, json, math, threading
from lib import logger, variables

class BinanceInteractor(logger.LogObject):
    def __init__(self, net):
        super().__init__('BinanceInteractor')
        self.variables = variables.Variables(net)
        self.w3 = web3.Web3(web3.HTTPProvider(self.variables.getSetting('bsc_url')))
        self.walletAdr = self.variables.getSetting('account_address')
        self.walletPk = self.variables.getSetting('account_pk')
        self.wallet = self.w3.eth.account.privateKeyToAccount(self.walletPk)
        self.runningOperations = [] # container of threads
        
    def address(self, address):
        return self.w3.toChecksumAddress(address)
    
    def wei(self, amt):
        return self.w3.toWei(amt, 'wei')

    def gwei(self, amt):
        return self.w3.toWei(amt, 'gwei')
    
    def ether(self, amt):
        return self.w3.toWei(amt, 'ether')
    
    def hex(self, str):
        return self.w3.toHex(str)
    
    def sleep(self):
        time.sleep(self.variables.getSetting('tx_delay'))

    def fromEthers(self, amt):
        return self.w3.fromWei(amt, 'ether')

    def fromWei(self, amt):
        return self.w3.fromWei(amt, 'wei')

    def createContract(self, abiID, address):
        self.logArgs(abiID, address)
        contractAbi = self.variables.getSetting(abiID)
        contract = self.w3.eth.contract(address=self.address(address), abi=contractAbi)
        return contract
    
    def createTransaction(self, smartContractFunc, txParams=None, increaseNonce=False):
        self.logArgs(smartContractFunc, txParams, increaseNonce)
        gasPrice = self.variables.getSetting('gas_price')
        gasPrice = self.gwei(gasPrice)
        gasAmount = self.variables.getSetting('gas_amount')
        nonce = self.w3.eth.getTransactionCount(self.address(self.walletAdr))
        params = {
            'nonce': nonce,
            'gasPrice': gasPrice,
            'from': self.address(self.walletAdr)
        }
        if addGas:
            params.update({'gas': gasAmount})
        if txParams is not None:
            params.update(txParams)
        self.log('Building TX', 'tx')
        tx = smartContractFunc.buildTransaction(params)
        self.log('Signing TX', 'tx')
        signedTx = self.w3.eth.account.sign_transaction(tx, self.walletPk)
        try:
            self.log('Sending TX', 'tx')
            txHash = self.w3.eth.send_raw_transaction(signedTx.rawTransaction)
            txHash = self.w3.toHex(txHash)
            self.log(txHash, 'txHash')
            return txHash
        except Exception as e:
            self.log('exception creating tx '+str(e), 'error')
            return None

    def getCoinInfo(self, token):
        self.logArgs(token)
        tokenContract = self.createContract('erc20_abi', token)
        name = tokenContract.functions.name().call()
        symbol = tokenContract.functions.symbol().call()
        balance = tokenContract.functions.balanceOf(self.walletAdr).call()
        rawBalance = balance
        balance = str(self.fromEthers(balance))
        balance = balance[:balance.index('.') + 3]
        ret = {'name': name, 'symbol': symbol, 'balance': balance, 'rawBalance': rawBalance}
        self.log(json.dumps(ret), 'coin info')
        return ret
    
    def approve(self, abiId, contractAddress, spender, amount):
        self.logArgs(abiId, contractAddress, spender, amount)
        approvalContract = self.createContract(abiId, contractAddress)
        approveCall = approvalContract.functions.approve(self.address(spender), amount)
        return self.createTransaction(approveCall)

    def getAmountsOut(self, amount, token1, token2):
        self.logArgs(amount, token1, token2)
        pancakeRouterContract = self.createContract('pancake_router_abi', self.variables.getSetting('ps_router_address'))
        getAmountsCall = pancakeRouterContract.functions.getAmountsOut(amount, [self.address(token1), self.address(token2)]).call()
        return getAmountsCall
    
    def swapExactETHForTokens(self, amount1, amount2, token1, token2):
        self.logArgs(amount1, token1, amount2, token2)
        pancakeRouterContract = self.createContract('pancake_router_abi', self.variables.getSetting('ps_router_address'))
        swapCall = pancakeRouterContract.functions.swapExactETHForTokens(
            amount2,
            [self.address(token1), self.address(token2)],
            self.address(self.walletAdr),
            (int(time.time()) + 10000000) 
        )
        extraParams = {
            'value': amount1
        }
        return self.createTransaction(swapCall, extraParams, False)
    
    def swapExactTokensForETH(self, amount1, token1, token2, amount2=0):
        self.logArgs(amount1, token1, token2)
        pancakeRouterContract = self.createContract('pancake_router_abi', self.variables.getSetting('ps_router_address'))
        swapCall = pancakeRouterContract.functions.swapExactTokensForETH(
            amount1,
            amount2,
            [self.address(token1), self.address(token2)],
            self.address(self.walletAdr),
            (int(time.time()) + 10000000) 
        )
        return self.createTransaction(swapCall, None, True)

    def sellToken(self, token, sellAmount):
        self.logArgs(token, sellAmount)
        self.log('Selling token '+token, 'sell')
        tokenToSellInfo = self.getCoinInfo(token)
        sellInEthers = self.ether(str(sellAmount))
        wbnbAdr = self.variables.getSetting('wbnb_address')
        if tokenToSellInfo['rawBalance'] < sellInEthers:
            self.log('Insufficent Balance on '+token, 'user-fail')
            return None
        else:
            approveHash = self.approve('basic_sell_abi', token, self.variables.getSetting('ps_router_address'), tokenToSellInfo['rawBalance'])
            if approveHash is not None:
                self.sleep()
                self.log('Approval hash '+approveHash, 'approved')
                swapHash = self.swapExactTokensForETH(sellInEthers, token, wbnbAdr)
                if swapHash is not None:
                    self.log('Swap hash '+swapHash, 'swapped')
                    return {
                        'swapHash': swapHash,
                        'approvalHash': approveHash,
                        'sellAmount': sellInEthers
                    }
                else:
                    self.log('Swap hash fail', 'fail')
                    return None
            else:
                self.log('Approval hash fail', 'fail')
                return None

    def buyToken(self, token):
        self.logArgs(token)
        self.log('Buying token '+token, 'buy')
        bnbBalance = self.w3.eth.getBalance(self.walletAdr)
        bnbAmount = self.ether(str(self.variables.getSetting('wbnb_buy_amount')))
        wbnbAdr = self.variables.getSetting('wbnb_address')
        approveHash = self.approve('basic_sell_abi', wbnbAdr, self.variables.getSetting('ps_router_address'), bnbBalance)
        if approveHash is not None:
            self.sleep()
            self.log('Approval hash '+approveHash, 'approved')
            amountsOut = self.getAmountsOut(bnbAmount, wbnbAdr, token)
            swapHash = self.swapExactETHForTokens(bnbAmount, math.floor(amountsOut[1] * 0.75), wbnbAdr, token)
            if swapHash is not None:
                self.log('Swap hash '+swapHash, 'swapped')
                self.kill_dead_operations()
                return {
                    'swapHash': swapHash,
                    'approveHash': approveHash,
                    'amounts': amountsOut
                }
            else:
                self.log('Swap hash fail', 'fail')
                self.kill_dead_operations()
                return None
        else:
            self.log('Approval hash fail', 'fail')
            self.kill_dead_operations()
            return None

    def sellToken(self, token, sellAmount):
        self.log('Selling token '+token, 'sell')
        tokenInfo = self.getCoinInfo(token)
        sellEthers = self.ether(sellAmount)
        tokenBalance = tokenInfo['rawBalance']
        approveHash = self.approve('basic_sell_abi', token, self.variables.getSetting('ps_router_address'), tokenBalance)
        if approveHash is not None:
            self.sleep()
            self.log('Approval hash '+approveHash, 'approved')
            swapHash = self.swapExactTokensForETH(sellEthers, token, self.variables.getSetting('wbnb_address'))
            if swapHash is not None:
                self.log('Swap hash '+swapHash, 'swapped')
                self.kill_dead_operations()
                return {
                    'swapHash': swapHash,
                    'approveHash': approveHash,
                    'amounts': [sellEthers]
                }
            else:
                self.log('Swap hash fail', 'fail')
                self.kill_dead_operations()
                return None
        else:
            self.log('Approval hash fail', 'fail')
            self.kill_dead_operations()
            return None
            
    def kill_dead_operations(self):
        removeIds = []
        for operation in self.runningOperations:
            thread = operation['operation']
            if not thread.is_alive():
                removeIds.append(operation['id'])

        for opId in removeIds:
            foundIndex = -1
            for i in range(0, len(self.runningOperations)):
                if self.runningOperations[i]['id'] == opId:
                    foundIndex = i
            if foundIndex != -1:
                self.runningOperations.pop(foundIndex)


    def start_buy(self, token):
        buyThread = threading.Thread(target=self.buyToken, args=(token,))
        self.runningOperations.append({
            'id': len(self.runningOperations),
            'operation': buyThread
        })
        buyThread.start()

    def start_sell(self, token, sellAmount):
        sellThread = threading.Thread(target=self.sellToken, args=(token,sellAmount,))
        self.runningOperations.append({
            'id': len(self.runningOperations),
            'operation': sellThread
        })
        sellThread.start()

File: v2_manager/lib/BinanceInteractions.js
const web3 = require('web3')
var Tx = require('ethereumjs-tx')
const Variables = require('./Variables.js')
const LogObject = require('./LoggingObject.js')

class BinanceInteractor extends LogObject{
    constructor(net) {
        super('BinanceInteractor')
        this.Variables = new Variables(net);
        this.W3 = new web3(this.Variables.getSetting('bsc_url'))
        this.W3.eth.defaultAccount = this.W3.utils.toChecksumAddress(this.Variables.getSetting('account_address'))
        if(this.W3 === undefined) {
            throw new Error('Error initializing web3')
        }
        return this;
    }

    /** Conversion Methods */
    address(address) {
        return this.W3.utils.toChecksumAddress(address)
    }

    wei(amt) {
        return this.W3.utils.toWei(amt, 'wei')
    }

    gwei(amt) {
        return this.W3.utils.toWei(amt, 'gwei')
    }

    ether(amt) {
        return this.W3.utils.toWei(amt, 'ether')
    }

    hex(str) {
        return this.W3.utils.toHex(str)
    }

    accountObj() {
        if(this.accountObject === undefined) {
            this.accountObject = this.W3.eth.accounts.privateKeyToAccount(this.Variables.getSetting('account_pk'))
        }
        return this.accountObject
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
      }

    /**
     * Returns the nice value of the amount of tokens in ether
     * @param {amt} String Amount of Tokens in Ether 
     * @returns 
     */
    fromEthers(amt) {
        return this.W3.utils.fromWei(amt, 'ethers')
    }

    fromWei(amt) {
        return this.W3.utils.fromWei(amt)
    }

    signAndSendTx(txObj, cb) {
        const tx = new Tx(txObj)
        this.log(JSON.stringify(txObj, null, 2), 'tx')
        tx.sign(Buffer.from(this.Variables.getSetting('account_pk'), 'hex'))

        const txHash = tx.serialize()
        const hashString = '0x' + txHash.toString('hex')
        return this.W3.eth.sendSignedTransaction(hashString, (err, txHash) => {
           return cb(err, txHash)
        })
    }

    smartContractInterAct(contract_data, interactionCallBack) {
        let txObj = {
            gasPrice: this.hex(this.gwei(this.Variables.getSetting('gas_price'))),
            gas: this.hex(this.Variables.getSetting('gas_amount')),
            from: this.address(this.Variables.getSetting('account_address')),
            data: contract_data,
            nonce: this.hex((nonce + 1).toString()),
            to: contract_address
        }
        return this.signAndSendTx(txObj, interactionCallBack.bind(this))
    }

    /** Smart Contract Methods */
    /**
     * Amount will be in the decimals of the token
     */
    approve(contract_abi_id, contract_address, spender, amount) {        
        this.logArgs(contract_abi_id, contract_address, spender, amount)
        let tokenContract = this.createContract(contract_abi_id, contract_address)

        return this.W3.eth.getTransactionCount(this.address(this.Variables.getSetting('account_address')))
        .then((nonce) => {
            let approve_data = tokenContract.methods.approve(this.address(spender), amount).encodeABI()
            
            const approveCallBack = function (err, txHash) {
                if(err) {
                    this.log('error: '+err, 'error')
                    throw new Error('Error approving spender')
                } else {
                    return txHash
                }
            }
            return this.smartContractInterAct(approve_data, approveCallBack)
        })        
    }

    getAmountsOutPancakeRouterWBNB(amountIn, token) {
        return new Promise(() => this.getAmountsOutWBNB('pancake_router_abi', this.Variables.getSetting('ps_router_address'), amountIn, token))
    }

    getAmountsOutPancakeFactoryWBNB(amountIn, token) {
        return new Promise(() => this.getAmountsOutWBNB('pancake_factory_abi', this.Variables.getSetting('ps_factory_address'), amountIn, token))
    }

    getAmountsOutWBNB(contract_abi_id, contract_address, amountIn, token) {
        this.logArgs(contract_abi_id, contract_address, amountIn, token)
        let tokenContract = this.createContract(contract_abi_id, contract_address)
        let amountArray = tokenContract.methods.getAmountsOutWBNB(amountIn, [this.address(this.Variables.getSetting('wbnb_address')), this.address(token)]).call()
        let amounts = {
            'wbnb': amountArray[0],
            'token': amountArray[1]
        }
        this.log('Amounts out :'+JSON.stringify(amounts, null, 2))
        return amounts
    }

    approvePancakeRouterWBNB(amount) {
        return new Promise(() => this.approve('basic_sell_abi', this.Variables.getSetting('wbnb_address'), this.Variables.getSetting('ps_router_address'), amount));
    }

    approvePancakeFactoryWBNB(amount) {
        return new Promise(() => this.approve('basic_sell_abi', this.Variables.getSetting('wbnb_address'), this.Variables.getSetting('pancakeswap_factory_abi'), amount));
    }

    createContract(abiID, address) {
        this.logArgs(abiID, address)
        let jsonABI = this.Variables.getSetting(abiID)
        if(jsonABI === undefined) {
            throw new Error('Contract not found in settings')
            return
        }
        let contract = new this.W3.eth.Contract(jsonABI, this.address(address))
        if(contract === undefined) {
            throw new Error('Error initalizing contract')
            return
        }
        return contract
    }

    swapExactTokenForWBNB(amountWBNB, amountToken) {
        return
        let gasPrice = this.gwei(this.this.Variables.getSetting('gas_price'));
        let routerContract = this.createContract('pancake_router_abi', this.Variables.getSetting('ps_router_address'))
        let swapCall = routerContract.methods.swapExactETHForTokens(
            amountToken,
            [ this.address(this.Variables.getSetting('wbnb_address')), this.address(token)],
            this.address(this.Variables.getSetting('account_address')), 
            (int(time.time()) + 10000000)
        )
        let gasAmount = this.Variables.getSetting('gas_amount')
        let nonce = this.W3.getTransactionCount(this.Variables.getSetting('account_address'))
        let swapTx = swapCall.buildTransaction({
            'from': this.Variables.getSetting('account_address'),
            'value': amountWBNB,
            'gas': gasAmount,
            'gasPrice': gasPrice,
            'nonce': nonce
        })
        this.log('Swapping wbnb for '+token)
        let signedTx = this.accountObj().sign_transaction(swapTx)
        try {
            let transactionSwap = this.W3.eth.sendRawTransaction(signedTx.rawTransaction)
            let transactionSwapHash = this.hex(transactionSwap)
            this.log('Swap Transaction Hash '+transactionSwapHash, 'hash')
            return transactionSwapHash
        } catch (e) {
            this.log(e.toString(), 'swap fail')
            return -1
        }
    }

    swapExactWBNBForToken(amountToken) {
        let gasPrice = this.gwei(this.this.Variables.getSetting('gas_price'));
        let routerContract = this.createContract('pancake_router_abi', this.Variables.getSetting('ps_router_address'))
        let swapCall = routerContract.methods.swapExactTokensForETH(
            amountToken, 0
            [ this.address(token), this.address(this.Variables.getSetting('wbnb_address'))],
            this.address(this.Variables.getSetting('account_address')), 
            (int(time.time()) + 10000000)
        )
        let nonce = this.W3.getTransactionCount(this.Variables.getSetting('account_address'))
        let swapTx = swapCall.buildTransaction({
            'from': this.Variables.getSetting('account_address'),
            'gasPrice': gasPrice,
            'nonce': nonce
        })
        this.log('Swapping wbnb for '+token)
        let signedTx = this.accountObj().sign_transaction(swapTx)
        try {
            let transactionSwap = this.W3.eth.sendRawTransaction(signedTx.rawTransaction)
            let transactionSwapHash = this.hex(transactionSwap)
            this.log('Swap Transaction Hash '+transactionSwapHash, 'hash')
            return transactionSwapHash
        } catch (e) {
            this.log(e.toString(), 'swap fail')
            return -1
        }
    }

    buyWithSlippage(token) {        
        this.logArgs(token)
        let bnbAmt = this.Variables.getSetting('wbnb_buy_amt').toString()
        this.log('Attempting to buy '+bnbAmt+ ' of token '+token, 'buy')
        this.log('Going to approve router first')
        return this.approvePancakeRouterWBNB(this.ether(bnbAmt)).then((hash) => {
            this.log('approve hash '+hash, 'hash')
            let amounts = this.getAmountsOutPancakeRouterWBNB(this.ether(bnbAmt), token)
            let swapHash = this.swapExactTokenForWBNB(amounts['wbnb'], amounts['token'] * this.Variables.getSetting('slippage'))
            return { approvalHash: hash, swapHash, ...amounts};
        })
    }

    getTokenBalance(token) {
        let tokenContract = this.createContract('balance_abi', token)
        let balanceCall = tokenContract.methods.balanceOf(this.address(this.Variables.getSetting('account_address'))).call()
        let balanceValues = {
            'ethers': balanceCall.toString(),
            'token': this.fromWei(balanceCall.toString()).toString()
        }
        return balanceValues
    }

    async sellToken(token, amountToSell) {
        this.logArgs(token, amountToSell)
        // check token balance
        let balance = this.getTokenBalance(token)
        if(amountToSell > balance['token']) {
            throw new Error('Amount to sell exceeds token balance')
        }
        this.log('Selling '+amountToSell+' of '+token)
        this.log('Approving the balance')
        let approvalHash = this.approvePancakeRouterWBNB(balance)
        await this.sleep(5)
        let swapHash = this.swapExactWBNBForToken(amountToSell.toString())
        return { approvalHash, swapHash, amountToSell }
    }

    watchToken(token, initalAmountToken, initialAmountBnb, moon='false') {
        this.logArgs(token, initalAmountToken, initialAmountBnb, moon)
        // will spawn an interval to get the balance
        if(this.watchers === undefined) {
            this.watchers = [];
        }
        let startingBNB = initialAmountBnb
        let targetBNB = startingBNB * this.Variables.getSetting('sell_target_percent')
        if(moon) {
            targetBNB = startingBNB / this.Variables.getSetting('sell_target_percent') * this.Variables.getSetting('moon_percent')
        }
        global['watch'+token] = initalAmountToken
        let watchInterval = setInterval( function(token, initalAmountToken, targetBNB, moon) {
            let amountsOut = this.getAmountsOutPancakeFactoryWBNB(initalAmountToken, token)
            let currentBNB = amountsOut['wbnb']
            if(global['watch'+token] > targetBNB) { 
                // we have surpassed the amount to sell
                if(moon) {
                    // handles the sell all for moon
                    sellToken(token, currentBNB)
                    global['watch'+token] = undefined
                } else {
                    sellToken(token, currentBNB * this.Variables.getSetting('sell_amount'))
                    // will need to handle moon watching as well
                    watchToken(token, initalAmountToken * this.Variables.getSetting('keep_amount'), currentBNB * this.Variables.getSetting('keep_amount'), moon=true)
                }
                clearInterval(this)
            } else {                
                global['watch'+token] = currentBNB;
            }
        }.bind(this), this.Variables.getSetting('watcher_interval'))
    }

    handlePairEvent(event) {
        let pairEvent = this.W3.utils.toJSON(event)
        console.log(pairEvent)
        let { token0, token1 } = pairEvent['args']
    }

    snipe(token) {
        this.logArgs(token)
        let uniswapContract = this.createContract('uniswap_sniper_abi', this.Variables.getSetting('us_sniper_address'))
        let tokenAddress = this.address(token)
        global['snipe'+token+'_event_filter'] = uniswapContract.events.PairCreated.createFilter(fromBlock='lastest')
        global['snipe'+token+'_event_filter_call'] = function () {
            let uniswapContract = this.createContract('uniswap_sniper_abi', this.Variables.getSetting('us_sniper_address'))
            global['snipe'+token+'_event_filter'] = uniswapContract.events.PairCreated.createFilter(fromBlock='lastest')
        }.bind(this)
        setInterval( () => {
            try {
                let filter = global['snipe'+token+'_event_filter']
                let newEntries = filter.get_new_entries()
                newEntries.forEach(event => {
                    this.handlePairEvent(event)
                }); 
            } catch (e) {   
                let exceptionMessage = e.toString()
                if(exceptionMessage.indexOf('filter not found') != -1) {
                    global['snipe'+token+'_event_filter_call']();
                } else {
                    this.log('Sniper exception '+e, 'exception')
                }
            }           
        }, this.Variables.getSetting('sniper_interval'));
        return true
    }
}

module.exports = BinanceInteractor;